### PR TITLE
ParticleSpecies: Read to pandas.DataFrame

### PR DESCRIPTION
**Pull request by ax3l**
Add a method tha reads a particle species into a `pandas.DataFrame`.

### To Do

- [x] rebase after #924 was merged

### CODE REVIEW

File: src/ParticleSpecies.cpp [hunk 1]
```diff
@@ -116,6 +118,31 @@ namespace
     }
 }
 
+void
+ParticleSpecies::seriesFlush()
+{
+    /** should look correctly somehow like this
```
Comment (ax3l) [inline src/ParticleSpecies.cpp:124]: @franzpoeschel Do you have an idea how we can implement this safely here? :)

File: src/ParticleSpecies.cpp [hunk 2]
```diff
@@ -116,6 +118,31 @@ namespace
     }
 }
 
+void
+ParticleSpecies::seriesFlush()
```
Comment (ax3l) [inline src/ParticleSpecies.cpp]: @franzpoeschel Do you have an idea how we can implement this method safely here? :)
If we find a general solution, we could later on also add this to a more general class like `Writable`, I think

I also tried looping through `parent->attributable` until I reach it's root via a `nullptr` but that did not lead anywhere yet

The commented attempts segfault, so I guess I can either not call `flush()` or `flush_impl()` like this or I did not hit the right object depth.

Comment (franzpoeschel) [reply src/ParticleSpecies.cpp]: [Little hint](https://github.com/openPMD/openPMD-api/blob/242cb1dbf5c508f6f8da3669f48fe26bbfc35b37/src/backend/Attributable.cpp#L151) ;) 
Needed to do the same trick for #901

I'm not quite sure though why your implementation didn't work, I don't see what I did different.

Comment (ax3l) [reply src/ParticleSpecies.cpp]: Ah perfect, I was close. Will try that construct, let's see :)

Comment (ax3l) [reply src/ParticleSpecies.cpp]: That works! :) Thank you :sparkles:

File: src/binding/python/openpmd_api/DataFrame.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import numpy as np
+try:
+    import pandas as pd
+    found_pandas = True
+except ImportError:
+    print("pandas NOT found. Install pandas for DataFrame support.")
+    found_pandas = False
+
+
+def particles_to_dataframe(particle_species, slice=None):
```
Comment (ax3l) [inline src/binding/python/openpmd_api/DataFrame.py:16]: Potential extra arguments:
- `records = iterable`: a list of strings or regex expressions to load only specific records of the species
- maybe an inverse parameter to deselect specific records from loading

File: examples/11_particle_dataframe.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+Copyright 2021 openPMD contributors
+Authors: Axel Huebl
+License: LGPLv3+
+"""
+import openpmd_api as io
+import numpy as np
+
+
+if __name__ == "__main__":
```
Comment (ax3l) [inline examples/11_particle_dataframe.py]: The main check here is actually not needed / not pretty.

Instead, we could add a guard here and execute this only if pandas was found

File: src/binding/python/openpmd_api/DataFrame.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+Copyright 2021 openPMD contributors
+Authors: Axel Huebl
+License: LGPLv3+
+"""
+import numpy as np
+try:
+    import pandas as pd
+    found_pandas = True
+except ImportError:
+    print("pandas NOT found. Install pandas for DataFrame support.")
```
Comment (dmitry-ganyushin) [inline src/binding/python/openpmd_api/DataFrame.py]: Do you have separation between critical errors and warnings? Why not to raise a custom exception here

File: examples/11_particle_dataframe.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env python
```
Comment (dmitry-ganyushin) [inline examples/11_particle_dataframe.py]: Is it python 2 or python3?

File: src/binding/python/openpmd_api/DataFrame.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        for rc_name, rc in record.items():
+            if record.scalar:
+                column_name = record_name
+            else:
+                column_name = record_name + "_" + rc_name
+            columns[column_name] = rc[slice]
+            particle_species.series_flush()
+            columns[column_name] *= rc.unit_SI
+
+    return pd.DataFrame(columns)
```
Comment (dmitry-ganyushin) [inline src/binding/python/openpmd_api/DataFrame.py:64]: How large are particle species data? It looks like panda lib has capability to work with tables not fitting into memory.

**Review by dmitry-ganyushin — APPROVED** [review]
Ok as a frist version, but see comments.

File: src/binding/python/openpmd_api/DataFrame.py (hunk 5 above)
Comment (ax3l) [reply src/binding/python/openpmd_api/DataFrame.py]: At this point I want to be able to not cause issues if pandas is not available but the function in this module is never called. I could also do this quietly, though, no reason to print here at all (throw is below).

File: examples/11_particle_dataframe.py (hunk 6 above)
Comment (ax3l) [reply examples/11_particle_dataframe.py]: All python3, could specify here but that is ensured anyway via packaging logic.

File: src/binding/python/openpmd_api/DataFrame.py (hunk 7 above)
Comment (ax3l) [reply src/binding/python/openpmd_api/DataFrame.py:64]: Particle data can be anything between 0 and 1e13 as of today.

Can you point me to information on how to handle out-of-core data in pandas?

File: examples/11_particle_dataframe.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+
+s = io.Series("../samples/git-sample/data%T.h5", io.Access.read_only)
+electrons = s.iterations[400].particles["electrons"]
+
+# all particles
+df = electrons.to_df()
+print(df)
+
+# only first 100 particles
```
Comment (ax3l) [inline examples/11_particle_dataframe.py:27]: extend example:
```
l_chunks = electrons.available_chunks()
```

**Pull request merged**